### PR TITLE
[RFC] vim-patch: 7.4.1259

### PR DESCRIPTION
**Pull request by KillTheMule**
I'll port https://github.com/vim/vim/commit/3fc3e14282c182c046d1335f3d576bc0eeb605c5, very easy. Just now, working on making tests/functional/legacy/increment_spec.lua into real lua.

### CODE REVIEW

**Comment by KillTheMule** [comment]
Conversion went without surprises. Is it okay to use \t as a textual representation of tab? I'm not sure how https://github.com/neovim/neovim/wiki/Unit-tests is supposed to be read in this case. I'll squash tomorrow and and the patch, if no one objects.

**Comment by KillTheMule** [comment]
Ok, done :) Many thanks @simonweil for helping with me screwing up git beyond recognition.

**Comment by KillTheMule** [comment]
Default passed *yay* Travis builds complain about missing luarocks... is there any way to restart without pushing a change?

**Comment by justinmk** [comment]
> Travis builds complain about missing luarocks

luarocks has been flakey today. Restarted the build.

**Comment by justinmk** [comment]
@KillTheMule Rebase?

**Comment by KillTheMule** [comment]
Done. Also snuk in marking 7.4.871 as merged.

**Comment by KillTheMule** [comment]
Travis failure looks unrelated, some kind of I/O error on osx.

File: test/functional/legacy/increment_spec.lua [hunk 1]
```diff
@@ -1,723 +1,742 @@
...
+  --     2)    Ctrl-A on visually selected "-10":
+  --     foobar-9
+  --     3)    Ctrl-A on visually selected "10":
+  --     foobar-11
+  --     4)    Ctrl-X on visually selected "-10"
+  --     foobar-11
+  --     5)    Ctrl-X on visually selected "10"
+  --     foobar-9
+  it('Test_visual_increment_01', function()
+    setline(1, repeater("foobaar-10",5))
```
Comment (jamessan) [inline test/functional/legacy/increment_spec.lua]: If I'm understanding helpers.call correclty, couldn't you replace `repeater("foobar-10",5)` with `call('repeat', {'foobar-10'}, 5)`?

**Comment by KillTheMule** [comment]
@jamessan Sorry can't answer to your line comment (pretty broken environment here), but thanks for the hint, I remember trying to make it work, but failed... your hint works though, so thanks a lot for the simplification.

File: test/functional/legacy/increment_spec.lua [hunk 2]
```diff
@@ -1,723 +1,731 @@
...
+  --     foobar-11
+  --     4)    Ctrl-X on visually selected "-10"
+  --     foobar-11
+  --     5)    Ctrl-X on visually selected "10"
+  --     foobar-9
+  it('Test_visual_increment_01', function()
+    setline(1, call('repeat', {'foobaar-10'}, 5))
+
+    cursor(1,1)
+    io.stderr:write(getline('.'))
```
Comment (jamessan) [inline test/functional/legacy/increment_spec.lua]: Was this being used for debug?  It doesn't seem relevant to the test.

**Comment by KillTheMule** [comment]
@jamessan Wow you're fast :) I repushed with some additional linting fixes. Thanks!

**Comment by KillTheMule** [comment]
QB failures should be fixed by https://github.com/neovim/neovim/pull/4718, travis failure looks very much unrelated (this PR only changes test files, and the test failing wasn't touched).

**Comment by KillTheMule** [comment]
From gitter discussion I've realized that the conversion of `increment_spec.lua` to "real lua" isn't really on the todo list... should I scrap that and simply add the test from the patch to the original file?

**Comment by justinmk** [comment]
@KillTheMule Up to you. If you feel confident about the converted test, I'm ok with it, though we should not spend time on it in the future.

**Comment by KillTheMule** [comment]
I will revert it, to keep things homogeneous and to not clutter the history with needless changes.

File: test/functional/legacy/increment_spec.lua [hunk 3]
```diff
@@ -720,4 +739,10 @@ describe('Ctrl-A/Ctrl-X on visual selections', function()
       eq({}, nvim.get_vvar('errors'))
     end)
   end
+
+  it('does not drop leading zeroes', function()
+    execute('set nrformats&vi') -- &vi makes Vim compatible
```
Comment (KillTheMule) [inline test/functional/legacy/increment_spec.lua]: Does this actually do anything? I carried it over, but cant see what it should do on nvim... `:h nrformats` wasn't helpful on vim as well, so if anyone could chime in here, that would be sweet :)

Comment (jamessan) [inline test/functional/legacy/increment_spec.lua]: That line isn't in the original test.  The original test has a SetUp function which performs `set nrformats&vim`.  This is needed because many of the tests modify `'nrformats'`, e.g. `set nrformats+=alpha`, so it needs to be reset to a known state before each test.

Since you're calling `clear()` before each lua test, that will achieve the same effect.

Comment (KillTheMule) [inline test/functional/legacy/increment_spec.lua]: It is in the original test, but there `set nrformats&vim` is called... but ok, if that's all  it does we can keep it :) But what's that syntax, can you explain that to me?

**Comment by KillTheMule** [comment]
Travis failure seems unrelated (but am I seeing that one a bit often? It's gcc on osx timing out on input_spec.lua... restarted anyways, let's see). QB has one missing tap plan, and one real failure, which I'm investigating now.

File: test/functional/legacy/increment_spec.lua (hunk 3 above)
Comment (jamessan) [inline test/functional/legacy/increment_spec.lua]: What I meant is it doesn't exist in the original test script in this function.  It's in the SetUp function, which is analogous to busted's before_each.

That syntax is explained in [`:help :set-default`](http://vimhelp.appspot.com/options.txt.html#:set-default)

Comment (KillTheMule) [inline test/functional/legacy/increment_spec.lua]: Ohhh "original" as in "original vim", I get it. Thanks! Also thanks for the help link, I got it now. I'm gonna keep it this way though, to keep the changes kinda minimal.

**Comment by KillTheMule** [comment]
Weellll of course _now_ everything passes. Bet the failures appear again after I've removed the debugging print...

**Comment by KillTheMule** [comment]
Ok, here's what we have: "test/functional/ui/input_spec.lua @ 61: feeding large chunks of input with <Paste> ok" on travis gcc on osx again, QB has a missing tap plan and `test/functional/autocmd/termclose_spec.lua @ 18` on freebsd-64. Just as before.

I'll give investigating QB another shot, but last time I put an "screen:snapshot_util()" in there to debug, everything passed...

**Comment by KillTheMule** [comment]
Why doesn't the QB start for this? Ping @jszakmeister could you have a look?

**Comment by jszakmeister** [comment]
@KillTheMule QB doesn't kick off right away like the other builds.  QB has to poll the project, which is does every 4 to 5 minutes.  There's not much I can do about that.

**Comment by KillTheMule** [comment]
Oh ok, I didn't know that, thanks! Seems it did run in the meantime :)

File: test/functional/terminal/tui_spec.lua [hunk 4]
```diff
@@ -151,7 +151,7 @@ describe('tui', function()
   it('can handle arbitrarily long bursts of input', function()
     execute('set ruler')
     local t = {}
-    for i = 1, 3000 do
+    for i = 2000, 3000 do
```
Comment (justinmk) [inline test/functional/terminal/tui_spec.lua]: this makes the output 60% shorter, so it is not quite the same test...

Comment (KillTheMule) [inline test/functional/terminal/tui_spec.lua]: Yeah of course, I just wanted to see if that's the issue here... doesn't change anything though!

Comment (justinmk) [inline test/functional/terminal/tui_spec.lua]: Something to try: make the 3 feed() calls into one.

     feed('i\027[200~' .. table.concat(t, '\n') .. '\027[201~')

I have a feeling this is a quirk in the existing bracketed-paste logic.

**Comment by KillTheMule** [comment]
Ok, so this is done now. The remaining issues are not related to this PR (which only updates a test anyways), and are fixed by https://github.com/neovim/neovim/pull/4757.

File: test/functional/terminal/tui_spec.lua (hunk 4 above)
Comment (jamessan) [inline test/functional/terminal/tui_spec.lua]: Thanks for figuring this out!  It definitely seems to be working now.  QB passed on both #4743 and #4737.
